### ISSUE TITLE
Out  of memory  sending 14gb to qingstor

### ISSUE THREAD

**Issue by 131** [ncw/rclone#2851]
#### What is the problem you are having with rclone?
rclone crashes with `fatal error: runtime: out of memory` error message
#### What is your rclone version (output from `rclone version`)
```
rclone v1.45           
- os/arch: linux/amd64 
- go version: go1.11.2 
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
linux - 64bit

####  Which cloud storage system are you using? (eg Google Drive)
swift to qingstor

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
```
rclone -vv copy repository:repository/files/8e/e/8ee02116ba8ea2314d65a67cd7eafc10 china:repository/files/8e/e/8ee02116ba8ea2314d65a67cd7eafc1

(swift to qingstor)
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

See https://gist.githubusercontent.com/131/0bd47404562f2841dc3390b577e3e568/raw/ce5205e85a7961c5937cefccadda79670a721586/traces_rclone_fuu_0

**Comment by ncw** [comment]
Thanks for reporting this.

Can you follow the instructions here: https://rclone.org/rc/#debugging-memory-use

And get a memory profile while rclone is running.  Remember to add the `--rc` parameter to your command line.

That should tell us where the memory is going - thanks.

**Comment by ncw** [comment]
Hmm... Looking at the code it uses 64M of memory for the buffer size.  The default concurrency is 4 so it is using 4 of these in the queue and it is uploading another 4 with another 1 waiting to go in the queue in the worst case.  That is 9 * 64M = 576M which is a bit large..

Assuming that `--transfers` is at the default 4 there can be 4 of these going on at once, so 4 * 576M of memory = 2.25GB of memory.

How many transfers were running at once when you took the memory dump?

How much memory (RSS) do you see rclone using.

Try reducing `--transfers` does that help it to go through - say `--transfers 1`?

How much memory do you have in your machine?

The qingstor backend could do with
- a control on the buffer size `--qingstor-chunk-size`
- a control on the concurrency `--qingstor-concurrency`
- a control on the cutoff between single and multi part `--qingstor-upload-cutoff`.

The MD5SUM calculations look wrong as they appear to rely on the chunks finishing in order.

A possible cause for the memory leak s clear in your backtrace above - there are 20 go routines `created by github.com/ncw/rclone/backend/qingstor.(*multiUploader).multiPartUpload`.  Since there is only one transfer active at the time of the crash these must have been leaked.  In your log we can also see low level retries - these must be leaking the connections.

Can you try this to see if it make it better?

https://beta.rclone.org/branch/v1.45-057-g25e50a74-fix-2851-qingstor-leak-beta/ (uploaded in 15-30 mins)

**Comment by 131** [comment]
Trying right now, with traces & stuffs. I'll keep you updated

**Comment by 131** [comment]
```
# in a tmux
watch -n 60 "go tool pprof -text http://localhost:5572/debug/pprof/heap > ta_\$(ls -la | wc -l)"
# in another pane
rclone --rc -vv copy repository:repository/files/ china:ivs-repository/files/ 2>&1 | tee traces
# with
cloudyks@dbx4:~$ rclone --version
rclone v1.45-057-g25e50a74-fix-2851-qingstor-leak-beta
- os/arch: linux/amd64
- go version: go1.11.4
# over 
2 cores / 4GB RAM

```

See traces : https://gist.githubusercontent.com/131/51a54910099872a125d95024170404e9/raw/ff618fd77c86e6d884b3bb4ef6d674b7e8663ba9/traces_rclone_fuu

And profiling : ```
File: rclone
Type: inuse_space
Time: Jan 3, 2019 at 9:52pm (CET)
Showing nodes accounting for 1.38GB, 99.14% of 1.39GB total
Dropped 63 nodes (cum <= 0.01GB)
      flat  flat%   sum%        cum   cum%
    1.38GB 99.14% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/backend/qingstor.(*uploader).nextReader
         0     0% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/backend/qingstor.(*Fs).Put
         0     0% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/backend/qingstor.(*Object).Update
         0     0% 99.14%     1.12GB 81.11%  github.com/ncw/rclone/backend/qingstor.(*multiUploader).multiPartUpload
         0     0% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/backend/qingstor.(*uploader).upload
         0     0% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/fs/operations.Copy
         0     0% 99.14%     1.38GB 99.14%  github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove

```

**Comment by 131** [comment]
Trying with `--transfert 1` in the meantime

**Comment by 131** [comment]
I'll gladly tests `--qingstor-concurrency 1`,  but i understand this is not an available flag

**Comment by ncw** [comment]
Here is a new version 

https://beta.rclone.org/branch/v1.45-058-g54208b02-fix-2851-qingstor-leak-beta/ (uploaded in 15-30 mins)

with some more parameters.  I've set these to the same defaults that S3 uses which may not be correct for qingstor.  The defaults should use a lot less memory.

Can you experiment with these?

#### --qingstor-upload-cutoff

Cutoff for switching to chunked upload

Any files larger than this will be uploaded in chunks of chunk_size.
The minimum is 0 and the maximum is 5GB.

- Config:      upload_cutoff
- Env Var:     RCLONE_QINGSTOR_UPLOAD_CUTOFF
- Type:        SizeSuffix
- Default:     200M

#### --qingstor-chunk-size

Chunk size to use for uploading.

When uploading files larger than upload_cutoff they will be uploaded
as multipart uploads using this chunk size.

Note that "--qingstor-upload-concurrency" chunks of this size are buffered
in memory per transfer.

If you are transferring large files over high speed links and you have
enough memory, then increasing this will speed up the transfers.

- Config:      chunk_size
- Env Var:     RCLONE_QINGSTOR_CHUNK_SIZE
- Type:        SizeSuffix
- Default:     4M

#### --qingstor-upload-concurrency

Concurrency for multipart uploads.

This is the number of chunks of the same file that are uploaded
concurrently.

If you are uploading small numbers of large file over high speed link
and these uploads do not fully utilize your bandwidth, then increasing
this may help to speed up the transfers.

- Config:      upload_concurrency
- Env Var:     RCLONE_QINGSTOR_UPLOAD_CONCURRENCY
- Type:        int
- Default:     4

**Comment by ncw** [comment]
> My machine has 4GB of RAM.
> 
> I don't understand how you can identify that memory is no longer an issue (the [full trace](https://gist.githubusercontent.com/131/51a54910099872a125d95024170404e9/raw/ff618fd77c86e6d884b3bb4ef6d674b7e8663ba9/traces_rclone_fuu) crashes with "fatal error: runtime: out of memory")

I can see from the trace that the go-routines which were holding onto memory aren't there any more....

The code is still using too much memory which I've fixed in the binary I posted above - the `058`.  You should find that uses a lot less memory.  If that one runs out of memory then there is a bug I haven't found yet!

**Comment by 131** [comment]
The Great Firewall God is not among us today (and i cannot transfer files anymore). I'll try again tomorrow...

**Comment by ncw** [comment]
How did it go?

**Comment by 131** [comment]
Crashing all the time, eating all the 4GB memory.

**Comment by ncw** [comment]
> Crashing all the time, eating all the 4GB memory.

What parameters are you using? Can you post a memory trace?

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.46

I think that you should be able to control the memory usage with the new parameters now.
